### PR TITLE
Add `ShellCommand` module for shell command stubs

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Follow-up to #17742, implementing @Bo98's idea https://github.com/Homebrew/brew/pull/17742#discussion_r1677300671. A RuboCop has also been added to make sure each stub has its counterpart.

### CODE REVIEW

File: Library/Homebrew/shell_command.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Homebrew
+  module ShellCommand
+    extend T::Helpers
+
+    requires_ancestor { T.class_of(AbstractCommand) }
```
Comment (Bo98) [inline Library/Homebrew/shell_command.rb:8]:
```suggestion
    requires_ancestor { AbstractCommand }
```

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/cmd/vendor-install.rb [hunk 2]
```diff
@@ -2,9 +2,12 @@
 # frozen_string_literal: true
 
 require "abstract_command"
+require "shell_command"
 
 module Homebrew
   module Cmd
+    include ShellCommand
+
     class VendorInstall < AbstractCommand
```
Comment (Bo98) [inline Library/Homebrew/cmd/vendor-install.rb:11]:
```suggestion
    class VendorInstall < AbstractCommand
      include ShellCommand
```

File: Library/Homebrew/dev-cmd/rubocop.rb [hunk 3]
```diff
@@ -2,18 +2,18 @@
...
 
 require "abstract_command"
+require "shell_command"
 
 module Homebrew
-  module Cmd
+  module DevCmd
+    include ShellCommand
+
     class Rubocop < AbstractCommand
```
Comment (Bo98) [inline Library/Homebrew/dev-cmd/rubocop.rb:11]:
```suggestion
    class Rubocop < AbstractCommand
      include ShellCommand
```

File: Library/Homebrew/shell_command.rb (hunk 1 above)
Comment (ZhongRuoyu) [reply Library/Homebrew/shell_command.rb]: Sadly now it fails with another error:

```
shell_command.rb:12: Method `dev_cmd?` does not exist on `T::Class[Homebrew::ShellCommand]` https://srb.help/7003
    12 |      sh_cmd_path = "#{self.class.dev_cmd? ? "dev-cmd" : "cmd"}/#{self.class.command_name}.sh"
                                          ^^^^^^^^
  Got `T::Class[Homebrew::ShellCommand]` originating from:
    shell_command.rb:12:
    12 |      sh_cmd_path = "#{self.class.dev_cmd? ? "dev-cmd" : "cmd"}/#{self.class.command_name}.sh"
                               ^^^^^^^^^^
shell_command.rb:12: Method `command_name` does not exist on `T::Class[Homebrew::ShellCommand]` https://srb.help/7003
    12 |      sh_cmd_path = "#{self.class.dev_cmd? ? "dev-cmd" : "cmd"}/#{self.class.command_name}.sh"
                                                                                     ^^^^^^^^^^^^
  Got `T::Class[Homebrew::ShellCommand]` originating from:
    shell_command.rb:12:
    12 |      sh_cmd_path = "#{self.class.dev_cmd? ? "dev-cmd" : "cmd"}/#{self.class.command_name}.sh"
                                                                          ^^^^^^^^^^
  There is a singleton class method with the same name:
    test/abstract_command_spec.rb:50: Defined here
    50 |            def self.command_name = "t-a-c"
                    ^^^^^^^^^^^^^^^^^^^^^
    Either:
    - use `.class` to call it, or
    - remove `self.` from its definition to make it an instance method
  Autocorrect: Use `-a` to autocorrect
    shell_command.rb:12: Insert `.class`
    12 |      sh_cmd_path = "#{self.class.dev_cmd? ? "dev-cmd" : "cmd"}/#{self.class.command_name}.sh"
                                                                                    ^
```

Any ideas?

Comment (Bo98) [reply Library/Homebrew/shell_command.rb]: Ah I think that covers the case where you can do `ShellCommand.new.run` which obviously we won't be but there isn't a way to state that with `requires_ancestor`.

Try add `T.bind(self, AbstractCommand)` to the top of `def run`.

Comment (ZhongRuoyu) [reply Library/Homebrew/shell_command.rb]: That worked. Thanks!

File: Library/Homebrew/cmd/--repository.rb [hunk 4]
```diff
@@ -2,10 +2,13 @@
 # frozen_string_literal: true
 
 require "abstract_command"
+require "shell_command"
 
 module Homebrew
   module Cmd
     class Repository < AbstractCommand
+      include ShellCommand
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/--repository.rb:10]: @ZhongRuoyu What's the thoughts on why `include ShellCommand` was a better fit here than `class Repository < ShellCommand`?

**Comment by MikeMcQuaid** [comment]
Nice work here, thanks @ZhongRuoyu!

I think a nice follow-up would be to rip out all the `#: `-handling logic for manpages now that we're not using them any more.

File: Library/Homebrew/cmd/--repository.rb (hunk 4 above)
Comment (Bo98) [reply Library/Homebrew/cmd/--repository.rb:10]: Haven't checked but the command system might detect ShellCommand as a command itself if it's a subclass of AbstractCommand.

**Pull request merged**